### ISSUE TITLE
How to use tfdbg in TFF

### ISSUE THREAD

**Issue by HaiQW** [tensorflow/federated#762]
Recently, I repeatedly  use the TFF framework to debug my federated algorithms, which is difficulty for me. 

I try to use `tfdbg`, but  can't explicitly find the session instance in TFF. How to use `tfdbg` to debug TFF ?

**Comment by jkr26** [comment]
Hi @HaiQW 

Thanks for your interest in TFF!

Generally, a recommended best practice in TFF is to debug your TensorFlow *first* before bringing it to TFF, since as you have noticed our runtime behavior is a little distant from the user; it sits at a layer below immediate user exposure. This is really a design choice, since TFF is designed to be deployable to production FL systems. When you wrap your Python functions with `@tff...._computation` decorators these functions are immediately traced, and a serialized representation of this function is given back to you. As the [Federated Core part 1 tutorial](https://www.tensorflow.org/federated/tutorials/custom_federated_algorithms_1#federated_computations) says, "when the Python interpreter encounters a function decorated with `tff.federated_computation`, it traces the statements in this function's body once (at definition time), and then constructs a serialized representation of the computation's logic for future use." It is this serialized representation that can be transformed and shipped around in a real distributed system, and it is towards this end that we have designed TFF in the first place.

The Python object that is given back to you is an instance of the `Computation` class, as you can check by calling Python `type` on this object. Long story short, when you invoke this object (IE call your function on some data), some runtime must interpret this serialized representation. For this purpose we have written two runtimes for TFF; one is the default [reference executor](https://github.com/tensorflow/federated/blob/22e80df3ae9ff5adb11595acf7de2a2f79cc1266/tensorflow_federated/python/core/impl/reference_executor.py), written for ease of reasoning about correctness and writing unit tests against, which really *does* use `Session.run` under the hood; see in particular lines 342-343 in the above file. You *could* instrument these `session.run` calls to use `tfdbg`, though since TFF is interpreting an object that is not your original graph or function, but a representation of this, the results may not be easy to understand. In particular, I am unaware of anybody trying this. Perhaps you will be the first!

Another option, if you are using the relatively new high-performance runtimes defined [here](https://github.com/tensorflow/federated/blob/22e80df3ae9ff5adb11595acf7de2a2f79cc1266/tensorflow_federated/python/core/impl/executor_stacks.py), is to instrument the underlying [EagerExecutor](https://github.com/tensorflow/federated/blob/22e80df3ae9ff5adb11595acf7de2a2f79cc1266/tensorflow_federated/python/core/impl/eager_executor.py). This executor executes all TensorFlow code in eager mode, however, so I am unsure how to use `tfdbg` in this situation; you may be better simply logging print statement at clever places inside the `EagerExecutor`.

In short, there are really three options here.
1. Recommended, and probably easiest, is to simply debug your TensorFlow with whatever tools you wish before decorating it with TFF.
2. My second option would be to use the high-performance (local) executor, and wire up print statements or other kinds of logging/profiling inside the `EagerExecutor`.
3. Finally, my last option would be actually using `tfdbg` inside of the reference executor. This is technically possible, but I think we would make few guarantees about its difficulty.

Hope this helps! And I hope option 1 will be sufficient!

**Comment by jkr26** [comment]
Pinging, is the answer above sufficient?

**Comment by HaiQW** [comment]
> Pinging, is the answer above sufficient?

Many thanks~
